pkg/controllers/nodeclass/status/ami_test.go

```diff
@@ -89,133 +91,426 @@ var _ = Describe("NodeClass AMI Status Controller", func() {
-					Name:         aws.String("test-ami-3"),
-					ImageId:      aws.String("ami-id-789"),
-					CreationDate: aws.String(time.Now().Add(2 * time.Minute).Format(time.RFC3339)),
-					Architecture: aws.String("arm64"),
-					Tags: []*ec2.Tag{
-						{Key: aws.String("Name"), Value: aws.String("test-ami-3")},
-						{Key: aws.String("foo"), Value: aws.String("bar")},
+					Name: "amd64-neuron",
+					ID:   "ami-amd64-neuron",
+					Requirements: []corev1.NodeSelectorRequirement{
+						{
+							Key:      corev1.LabelArchStable,
+							Operator: corev1.NodeSelectorOpIn,
+							Values:   []string{karpv1.ArchitectureAmd64},
+						},
+						{
+							Key:      v1.LabelInstanceAcceleratorCount,
+							Operator: corev1.NodeSelectorOpExists,
+						},
 					},
 				},
-			},
+				{
+					Name: "arm64-standard",
+					ID:   "ami-arm64-standard",
+					Requirements: []corev1.NodeSelectorRequirement{
+						{
+							Key:      corev1.LabelArchStable,
+							Operator: corev1.NodeSelectorOpIn,
+							Values:   []string{karpv1.ArchitectureArm64},
+						},
+						{
+							Key:      v1.LabelInstanceGPUCount,
+							Operator: corev1.NodeSelectorOpDoesNotExist,
+						},
+						{
+							Key:      v1.LabelInstanceAcceleratorCount,
+							Operator: corev1.NodeSelectorOpDoesNotExist,
+						},
+					},
+				},
+			}))
+			Expect(nodeClass.StatusConditions().IsTrue(v1.ConditionTypeAMIsReady)).To(BeTrue())
 		})
-		nodeClass.Spec.AMIFamily = lo.ToPtr(v1.AMIFamilyAL2)
-		nodeClass.Spec.AMISelectorTerms = []v1.AMISelectorTerm{{Alias: "al2@latest"}}
-		ExpectApplied(ctx, env.Client, nodeClass)
-		ExpectObjectReconciled(ctx, env.Client, statusController, nodeClass)
-		nodeClass = ExpectExists(ctx, env.Client, nodeClass)
-		Expect(len(nodeClass.Status.AMIs)).To(Equal(4))
-		Expect(nodeClass.Status.AMIs).To(ContainElements([]v1.AMI{
-			{
-				Name: "test-ami-3",
-				ID:   "ami-id-789",
-				Requirements: []corev1.NodeSelectorRequirement{
-					{
-						Key:      corev1.LabelArchStable,
-						Operator: corev1.NodeSelectorOpIn,
-						Values:   []string{karpv1.ArchitectureArm64},
+		It("Should resolve all AMIs with correct requirements for AL2", func() {
+			awsEnv.SSMAPI.Parameters = map[string]string{
+				fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2/recommended/image_id", k8sVersion):       "ami-amd64-standard",
+				fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-gpu/recommended/image_id", k8sVersion):   "ami-amd64-nvidia",
+				fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-arm64/recommended/image_id", k8sVersion): "ami-arm64-standard",
+			}
+			nodeClass.Spec.AMISelectorTerms = []v1.AMISelectorTerm{{Alias: "al2@latest"}}
+			ExpectApplied(ctx, env.Client, nodeClass)
+			ExpectObjectReconciled(ctx, env.Client, statusController, nodeClass)
+			nodeClass = ExpectExists(ctx, env.Client, nodeClass)
+
+			Expect(len(nodeClass.Status.AMIs)).To(Equal(4))
+			Expect(nodeClass.Status.AMIs).To(ContainElements([]v1.AMI{
```

Comment:
Do you really need to redefine it? Can't you just use the AMIs that you defined up above in a variable and reference it? Saves on code lines.

Reply:
When you say up above, are you referring to the expected AMIs in the status conditions from AL2023? In that particular example they're different since AL2023 has different AMIs for neuron and nvidia. Even if they were the same, I don't think we want to couple them in the test since it's really just happenstance. I.e. if the AMIs available for one family changed, that doesn't mean the AMIs for another would change in kind.